src/test/java/org/apache/maven/shared/utils/cli/CommandLineUtilsTest.java

```diff
@@ -168,4 +169,25 @@ private void assertCmdLineArgs( final String[] expected, final String cmdLine )
         assertEquals( expected.length, actual.length );
         assertEquals( Arrays.asList( expected ), Arrays.asList( actual ) );
     }
+
+    @Test
+    public void testChineseEncodingIssue()
```

Comment:
This test is completely pointless:

1. The `echo` will use the encoding supplied the entire system. There is no guarantee that GBK is the system encoding.
2. You never verify the output of th command to be what you expect.

What you require is an application that wroduces GBK *bytes* , those are read by Java with GBK into a `String` then you need to compared this value.

Reply:
As far as i am concerned, the parameter `commandLine` of `CommandLineUtils.executeCommandLineAsCallable` is used to create a `Process` object, which is a result of  `Runtime.getRuntime().execute()`. This `execute()` method uses different encoding depending on different system. 

Any idea of producing GBK bytes using `CommandLineUtils.executeCommandLineAsCallable`? Or i can modify the test to use system encoding rather than using GBK.

Reply:
By the way, this test will pass without this PR if system encoding is the same as the result of `Character.defaultCharset()`. So the influence of this fix may not be very obvious.

Reply:
> As far as i am concerned, the parameter commandLine of CommandLineUtils.executeCommandLineAsCallable is used to create a Process object, which is a result of Runtime.getRuntime().execute(). This execute() method uses different encoding depending on different system.

Why do you think so? It uses the same encoding as the surrounding Java process does. You cannot change this really on Windows, on Unix you can pass LC_ALL to the env.

> Any idea of producing GBK bytes using CommandLineUtils.executeCommandLineAsCallable? Or i can modify the test to use system encoding rather than using GBK.

You have two options:

1. Modify `file.encoding` and set back in the finally block. Implies you read the output stream. I don't exactly know whether tests can run in parallel in the same JVM, this could break other tests.
2. Write a simple Java program, put it in `src/test/java`, call the `.class` file with Java from within the test. It should use `System.out` as a byte-oriented stream which will write bytes according to GBK. Read those with the consumer and check when normalized back to UTF-16.

Reply:
> Write a simple Java program, put it in src/test/java, call the .class file with Java from within the test. It should use System.out as a byte-oriented stream which will write bytes according to GBK. Read those with the consumer and check when normalized back to UTF-16.

I tried producing gbk bytes with `System.out.println(new String("金色传说".getBytes(), "GBK")`. When comparing value in the cousumer, test passes on a windows-gbk platform but fails on a mac-utf-8 platform. Maybe it's not right to produce gbk bytes like that.

Reply:
That's wrong. You need to use System.out as an byte stream, not a char stream: `byte[] bytes = "...".getBytes(encoding)` then `System.out.write(bytes)`.